libraries/lib-wave-track/WaveClip.cpp

```diff
@@ -199,6 +254,82 @@ size_t WaveClip::GetAppendBufferLen() const
    return mSequences[0]->GetAppendBufferLen();
 }
 
+void WaveClip::OnProjectTempoChange(
+   const std::optional<double>& oldTempo, double newTempo)
+{
+   if (!mRawAudioTempo.has_value())
+      // When we have tempo detection ready (either by header-file
+      // read-up or signal analysis) we can use something smarter than that. In
+      // the meantime, use the tempo of the project when the clip is created as
+      // source tempo.
+      mRawAudioTempo = oldTempo.value_or(newTempo);
+
+   if (oldTempo.has_value())
+   {
+      const auto ratioChange = *oldTempo / newTempo;
+      mSequenceOffset *= ratioChange;
+      mTrimLeft *= ratioChange;
+      mTrimRight *= ratioChange;
+      mEnvelope->RescaleTimesBy(ratioChange);
+   }
+   mProjectTempo = newTempo;
+}
+
+void WaveClip::StretchLeftTo(double to)
+{
+   const auto pet = GetPlayEndTime();
+   if (to >= pet)
+      return;
+   const auto oldPlayDuration = pet - GetPlayStartTime();
+   const auto newPlayDuration = pet - to;
+   const auto ratioChange = newPlayDuration / oldPlayDuration;
+   mSequenceOffset = pet - (pet - mSequenceOffset) * ratioChange;
+   mTrimLeft *= ratioChange;
+   mTrimRight *= ratioChange;
+   mClipStretchRatio *= ratioChange;
+   mEnvelope->SetOffset(mSequenceOffset);
+   mEnvelope->RescaleTimesBy(ratioChange);
+   StretchCutLines(ratioChange);
+}
+
+void WaveClip::StretchRightTo(double to)
+{
+   const auto pst = GetPlayStartTime();
+   if (to <= pst)
+      return;
+   const auto oldPlayDuration = GetPlayEndTime() - pst;
+   const auto newPlayDuration = to - pst;
+   const auto ratioChange = newPlayDuration / oldPlayDuration;
+   mSequenceOffset = pst - mTrimLeft * ratioChange;
+   mTrimLeft *= ratioChange;
+   mTrimRight *= ratioChange;
+   mClipStretchRatio *= ratioChange;
+   mEnvelope->SetOffset(mSequenceOffset);
+   mEnvelope->RescaleTimesBy(ratioChange);
+   StretchCutLines(ratioChange);
+}
+
+void WaveClip::StretchCutLines(double ratioChange)
+{
+   for (const auto& cutline : mCutLines)
+   {
+      const auto startTime = cutline->GetPlayStartTime();
+      const auto newPlayStartTime = cutline->GetPlayStartTime() * ratioChange;
+      const auto newDuration = cutline->GetPlayDuration() * ratioChange;
+      cutline->ShiftBy(newPlayStartTime - startTime);
+      cutline->StretchRightTo(newPlayStartTime + newDuration);
```

Comment:
We both forgot something about cutlines last week:  they too can have envelopes and paste envelope points into the clip when they are expanded.

Did you test this possibility?

Just as, one of these commits is a one-line fix to rescale envelopes of the main clip -- so too here, I think you must remap any envelope points in the cutline.

Reply:
I coincidentally thought about this earlier today and already pushed a fix. Spot on, though.